### ISSUE TITLE
How to add rule to attached sheet without rerenders

### ISSUE THREAD

**Issue by zerkalica** [cssinjs/jss#657]
I try to delete many rules in attached sheet dinamically, but addRule calls renderer.insertRule and modify dom.

I try to detach sheet before adding/deleting, but detach removes existing styles from dom.

```
sheet.detach()
sheet.addRule(...)
sheet.deleteRule(...)
sheet.deleteRule(...)
sheet.attach()
```

May be expose some sheet.beginModify() / sheet.endModify() ?

Why i need it. I try to create dynamically updatable css without Update function values, moving all sheet updates to next requestAnimationFrame and using object memoization by JSON.stringify. This method needs to to utilize old rules in next frame, but simplifies react-jss wrappers and jss usage in components without perfomance penalty. Something like this:

```
function Component(props) {
  return <div class={{background: `${props.color}`}}>some</div>
}
```

**Comment by kof** [comment]
I completely fail to understand what are you actually trying to achieve. Please explain step by step or create a demo on  codesandbox

**Comment by kof** [comment]
The most important information is "why" would you want to do something like that is crucial, it seems like you are going to do things you probably shouldn't and you are basically need a guidance to achieve it in a different way.

**Comment by zerkalica** [comment]
I just want modify rules in attached sheet transactionally.

I use jss as stylesheet renderer for my experimental [reactive-di](https://github.com/zerkalica/reactive-di). It's an framework with dependency injection and reactivity, which use react and jss as backends, but tries to abstract userland code from this libraries.

For example, style in [TodoAppView](https://github.com/zerkalica/rdi-examples/blob/master/src/examples/todomvc/TodoAppView.js) just an object, wrapped in decorator theme, which memoizes jss object. This works fine, but decorator converts jss style object to classnames object. I want to try pass pure object to class property. Redefined createElement in jsx autoconverts it to string.

Compare this example with [react-jss](https://github.com/cssinjs/react-jss#example):

```ts
class Theme {
  colorPrimary: string = 'red'
}

function Button(
  {color, children},
  {theme}: {theme: Theme}
) {
  return <button class={{ background: theme.colorPrimary }}>
    <span class={{ background: props.color }}>
      {children}
    </span>
  </button>
}
```

May be i wrong, by it looks simpler:

1. Dependency injection gives type safety and replaces react contexts (they not typesafe) and can be used instead of ThemeProvider.
2. DI can be used instead of injectSheet to inject theme variables to context argument, and this feature can be used not only for themes.
3. DI gives abstraction from libraries, above example is only business logic, just specifications: jsx, jss-styles. I can configure di with any css-in-js or react-like backends without touching application code.
4. We can use jss object in-place in class property (autoconverted to string in custom createElement). Yes, slightly perfomance degraded, but simple usage.
5. In complex styles we can still extract jss object into separate object and use any memoization library (mobx, cellx, etc) to improve perfomance.
6. We don't need to use functions as props in jss-object for injecting variables

Look at the [emotion.sh](https://emotion.sh/#css-prop), it use magic css property and converts it to js via babel plugin:

```js
function SomeComponent (props) {
  // Create styles as if you're calling css and the class will be applied to the component
  return (<div css={`
    color: blue;
    font-size: ${props.fontSize}px;
 ....
}
```

I just want something like this, but independend from concrete css-in-js library.

**Comment by kof** [comment]
That turns out to be a long and unfocused discussion. You are going to need to change the class names generator to use hashes instead of id's.

Methods `deleteRule` and `addRule` should work in attached and detached state. If you find a specific bug, please file a specific issue with a small demo on codesandbox. 

Hope this helps.
